### PR TITLE
fix: use helm secrets view rather than helm secrets dec to decrypt

### LINKED ISSUE #193: Incompatible with latest helm-secrets
@jkroepke mentioned in https://github.com/roboll/helmfile/issues/1878#issuecomment-860043605 an upcoming change in helm-secrets which would break helmfile, which has since been released and breaks helmfile. I think the following code section needs to be updated to correspond to the newer helm-secrets behavior that was described in the linked comment: https://github.com/helmfile/helmfile/blob/9dca4fea5926c2ae4266e4ac1cad3b72ef9afed9/pkg/helmexec/exec.go#L282-L294

With the current code, an error occurs whenever using helmfile with helm-secrets and a custom `HELM_SECRETS_DEC_SUFFIX` and helm-secrets 3.8.2.

Repro: https://github.com/raxod502-plaid/helmfile-issue-1

In that repo with
* helm 3.3.4
* helm-secrets 3.8.2
* helmfile 0.144.0

Then if I run `helmfile template` it works:

```
% helmfile template
Decrypting secret /Users/rrosborough/temp/helmfile-repro/secrets.yaml
[helm-secrets] Decrypting /Users/rrosborough/temp/helmfile-repro/secrets.yaml

Building dependency release=hewwo, chart=hewwo
Templating release=hewwo, chart=hewwo
```

However if I use a custom decryption suffix it is broken:

```
% HELM_SECRETS_DEC_SUFFIX=.dec helmfile template
Decrypting secret /Users/rrosborough/temp/helmfile-repro/secrets.yaml
[helm-secrets] Decrypting /Users/rrosborough/temp/helmfile-repro/secrets.yaml

open /Users/rrosborough/temp/helmfile-repro/secrets.dec: no such file or directory
open /Users/rrosborough/temp/helmfile-repro/secrets.dec: no such file or directory
in ./helmfile.yaml: failed to read helmfile.yaml: failed loading environment secrets with 1 errors
```

Furthermore, the file `secrets.yaml.dec` is created in the repo and is not cleaned up. It seems that helmfile and helm-secrets are in disagreement about the name of the decrypted file.

Note that this worked back in helm-secrets 3.7.0-dev, but it's broken in 3.8.2 due to the aforementioned upstream update.

Using `HELM_SECRETS_DEC_SUFFIX` is required for our use case due to https://github.com/roboll/helmfile/issues/1878#issuecomment-860043605 to which there does not appear to be another workaround at present.

### PR DESCRIPTION
**Pull request by philomory**
Signed-off-by: Adam Gardner <adam.gardner@magicmemories.com>

Fixes #193

### CODE REVIEW

**Comment by philomory** [comment]
There were two test cases that failed; one was easy to fix: the test case was expecting `exec: helm secrets dec ...` to get logged, and that needed to change to `exec: helm secrets view ...`.

The other failed test case, I'm less sure about, because I'm honestly not sure what it's doing. It looks like it's expecting an error but not getting one. This is `Test_DecryptSecretWithGotmpl`.

**Comment by yxxhero** [comment]
@philomory Please fix the ci issue.

**Comment by raxod502-plaid** [comment]
I fixed the tests in https://github.com/helmfile/helmfile/pull/204, feel free to merge that into this PR and continue here if you like. Just wanted to help get the issue fixed since working around it is quite tricky.

**Comment by philomory** [comment]
@yxxhero I think this is ready for review again.

**Comment by yxxhero** [comment]
@mumoshu ping

**Comment by yxxhero** [comment]
@philomory Please sqush your commits.

File: pkg/helmexec/exec_test.go [hunk 1]
```diff
@@ -362,10 +362,7 @@ func Test_DecryptSecret(t *testing.T) {
 
 	_, err := helm.DecryptSecret(HelmContext{}, "secretName")
 	if err != nil {
-		if _, ok := err.(*os.PathError); ok {
```
Comment (yxxhero) [inline pkg/helmexec/exec_test.go:365]: why change?

**Comment by philomory** [comment]
> @philomory Please sqush your commits.

@yxxhero Do you want me to squash only my commits? Currently there are four commits in this PR: two from myself, and two from @raxod502-plaid . I can squash only my own commits (leaving one from myself and two from @raxod502-plaid ), but if I squash @raxod502-plaid's commits I think that will make it appear that I am the author of those commits, rather than him.

**Comment by yxxhero** [comment]
@philomory yes. do as you say. Thanks very much. My english not good.

**Comment by philomory** [comment]
Ok, I've squashed my two commits

File: pkg/helmexec/exec_test.go (hunk 1 above)
Comment (yxxhero) [reply pkg/helmexec/exec_test.go:365]: @philomory

Comment (raxod502-plaid) [reply pkg/helmexec/exec_test.go:365]: I changed that because the new implementation doesn't return a PathError.

Comment (raxod502-plaid) [reply pkg/helmexec/exec_test.go:365]: To be frank, I don't fully understand why the test was the way it was initially. I looked in the commit history but there was no explanation included with the PR that added it.

Comment (yxxhero) [reply pkg/helmexec/exec_test.go:365]: @philomory @raxod502-plaid Thanks very much.

File: pkg/helmexec/exec_test.go [hunk 2]
```diff
@@ -402,14 +401,27 @@ func Test_DecryptSecretWithGotmpl(t *testing.T) {
...
 
-	expected := fmt.Sprintf(`%s/%s.yaml.dec`, cwd, secretName)
-	if d := cmp.Diff(expected, decryptErr.(*os.PathError).Path); d != "" {
+	expected := fmt.Sprintf(`Preparing to decrypt secret %v/secretName.yaml.gotmpl
+Decrypting secret %s/secretName.yaml.gotmpl
+exec: helm --kube-context dev secrets view %s/secretName.yaml.gotmpl
+Decrypted %s/secretName.yaml.gotmpl into %s
+`, cwd, cwd, cwd, cwd, tmpFilePath)
+	if err != nil {
+		if _, ok := err.(*os.PathError); ok {
```
Comment (yxxhero) [inline pkg/helmexec/exec_test.go:419]: ok{}, empty? please fix it.

Comment (yxxhero) [reply pkg/helmexec/exec_test.go:419]: @philomory

Comment (philomory) [reply pkg/helmexec/exec_test.go]: @yxxhero Like in the previous test case, `os.PathError` here is no longer special, nor needs any special handling. I've removed that check entirely; if there is an error, we now treat it as such, no matter what type of error it was.

**Comment by mumoshu** [comment]
Hey! Thanks for the PR. This looks good in general. But one question- Don't we need to upgrade Dockerfiles? Our dockerfiles point to helm-secrets 3.5.0. I think it breaks since 3.8.0 but what's the minimum supported helm-secrets version after this logic change?

**Comment by philomory** [comment]
The `helm secrets view` subcommand has been part of `helm secrets` since v1.0. As far as I am aware nothing has changed with it recently. The immediate impetus for switching from `dec` to `view` was a change in he behavior of `dec`, but, we could have been using `view` the whole time.

**Comment by mumoshu** [comment]
@philomory Good to know! Thanks a lot for clarifying ☺️

**Comment by yxxhero** [comment]
@mumoshu approved this PR?

**Review by mumoshu — APPROVED** [review]
LGTM. Thanks a lot for your contribution @philomory!

**Review by yxxhero — APPROVED** [review]

**Pull request merged**